File: 238-product-of-array-except-self/product-of-array-except-self.kt
class Solution {
    fun productExceptSelf(nums: IntArray): IntArray {
        val size = nums.size
        val pre = IntArray(size)
        val post = IntArray(size)
        val result = IntArray(size)

        var prev = 1
        var i = 0
        while(i < nums.size) {
            pre[i] = nums[i] * prev
            prev = pre[i]
            i++
        }

        var j = size - 1
        post[j] = 1
        while(j > 0) {
            post[j-1] = post[j] * nums[j]
            j--
        }

        var k = 1
        result[0] = post[0]
        while(k < size){
            result[k] = post[k] * pre[k-1]
            k++
        }

        return result

    }


}

File: 238-product-of-array-except-self/product-of-array-except-self.go
func productExceptSelf(nums []int) []int {
	prefix := make([]int, len(nums))
	suffix := make([]int, len(nums))

	result := make([]int, len(nums))
	prefix_cur, suffix_cur := 1, 1

	for i, num := range nums {
		prefix[i] = prefix_cur
		prefix_cur *= num
	}

	for j := range nums {
		suffix[len(nums)-1-j] = suffix_cur
		suffix_cur *= nums[len(nums)-1-j]
	}

	for i := range prefix {
		result[i] = prefix[i] * suffix[i]
	}

	return result
}